src/endian_reader.rs

```diff
@@ -0,0 +1,428 @@
+    T: Deref<Target = [u8]> + Clone + Debug,
+{
+    type Endian = Endian;
+    type Offset = usize;
+
+    #[inline]
+    fn endian(&self) -> Endian {
+        self.endian
+    }
+
+    #[inline]
+    fn len(&self) -> usize {
+        self.range.len()
+    }
+
+    #[inline]
+    fn is_empty(&self) -> bool {
+        self.range.start == self.range.end
+    }
+
+    #[inline]
+    fn empty(&mut self) {
+        self.range.start = self.range.end;
+    }
+
+    #[inline]
+    fn truncate(&mut self, len: usize) -> Result<()> {
+        if self.len() < len {
+            Err(Error::UnexpectedEof)
+        } else {
+            self.range.start = self.range.end - len;
+            Ok(())
+        }
+    }
+
+    #[inline]
+    fn offset_from(&self, base: &EndianReader<Endian, T>) -> usize {
+        let base_ptr = base.bytes().as_ptr() as *const u8 as usize;
+        let ptr = self.bytes().as_ptr() as *const u8 as usize;
+        debug_assert!(base_ptr <= ptr);
+        debug_assert!(ptr + self.bytes().len() <= base_ptr + base.bytes().len());
+        ptr - base_ptr
+    }
+
+    #[inline]
+    fn find(&self, byte: u8) -> Result<usize> {
+        self.range.find(byte).ok_or(Error::UnexpectedEof)
+    }
+
+    #[inline]
+    fn skip(&mut self, len: usize) -> Result<()> {
+        if self.len() < len {
+            Err(Error::UnexpectedEof)
+        } else {
+            self.range.start += len;
+            Ok(())
+        }
+    }
+
+    #[inline]
+    fn split(&mut self, len: usize) -> Result<Self> {
+        if self.len() < len {
+            Err(Error::UnexpectedEof)
+        } else {
+            let mut r = self.clone();
+            r.range.end = r.range.start + len;
+            self.range.start += len;
+            Ok(r)
+        }
+    }
+
+    #[inline]
+    fn to_slice(&self) -> Result<Cow<[u8]>> {
+        Ok(self.bytes().into())
+    }
+
+    #[inline]
+    fn to_string(&self) -> Result<Cow<str>> {
+        match str::from_utf8(self.bytes()) {
+            Ok(s) => Ok(s.into()),
+            _ => Err(Error::BadUtf8),
+        }
+    }
+
+    #[inline]
+    fn to_string_lossy(&self) -> Result<Cow<str>> {
+        Ok(String::from_utf8_lossy(self.bytes()))
+    }
+
+    #[inline]
+    fn read_u8_array<A>(&mut self) -> Result<A>
+    where
+        A: Sized + Default + AsMut<[u8]>,
+    {
+        let len = mem::size_of::<A>();
+        let slice = self.range.read_slice(len)?;
+        let mut val = Default::default();
+        <A as AsMut<[u8]>>::as_mut(&mut val).clone_from_slice(slice);
+        Ok(val)
+    }
+
+    #[inline]
+    fn read_u8(&mut self) -> Result<u8> {
+        let slice = self.range.read_slice(1)?;
+        Ok(slice[0])
+    }
+
+    #[inline]
+    fn read_i8(&mut self) -> Result<i8> {
+        let slice = self.range.read_slice(1)?;
+        Ok(slice[0] as i8)
+    }
+
+    #[inline]
+    fn read_u16(&mut self) -> Result<u16> {
+        let slice = self.range.read_slice(2)?;
+        Ok(self.endian.read_u16(slice))
```

Comment:
If we can make it work, then totally. I think using `read_u8_array` would avoid the borrow issues I was running into, since it makes a copy. The copy is of a small size, and I expect llvm should also optimize it away, but verifying that requires digging into the disassembly.